code-collab-app/src/storage/peer_data.js

```diff
@@ -68,71 +109,150 @@ const peer_data = observable({
-                console.log(data);
             });
         }
+        else{
+            // Making a call to the tracker to get doc data
+            this.tracker.emit(GET_DOC_REQ, this.current_session_id);
+        }
     },
 
     //  listen for other peer's initial connection
     //  and their request for the doc data
     listen_for_req() {
         this.peer.on('connection', dataConnection => {
-            dataConnection.on('error', err => {
-                console.log("listen for req err: " + err);
-            });
+            // adding the newly connected peer to the list
+            dataConnection.on('open', () => {
+                this.session_peers.push(dataConnection.peer);
+                this.session_peers_conn.push(dataConnection);
 
-            dataConnection.on('data', data => {
-                dataConnection.send(this.doc_data);
+                dataConnection.on('error', error => {
+                    this.handle_error_from_peers(error);
+                });
+    
+                dataConnection.on('data', data => {
+                    this.handle_data_from_peers(data, dataConnection);
+                });
+    
+                // listening to errors on any data connections
+                dataConnection.on('close', () => {
+                    this.handle_peer_disconnect(dataConnection);
+                });                       
             });
         });
     },
 
-    upload_document(d) {
-        this.doc_data = d;
+    handle_data_from_peers(data, dataConnection){
+        // Received a get doc request
+        if(data.message_type && data.message_type === GET_DOC_REQ){
+            console.log('received a get doc data request from another peer, sending data...');
+            dataConnection.send({
+                message_type: GET_DOC_RES,
+                content: this.doc_data
+            });
+        }
+        // Received a get doc response
+        else if(data.message_type && data.message_type === GET_DOC_RES && data.content){
+            console.log('received doc data from another peer');
+            this.doc_data.push(...data.content);
+        }
+        else{
+            // TODO -- do something more with this data
+            console.log(data);
+        }
+    },
+
+    handle_error_from_peers(error){
+        console.log(error);
+    },
+
+    handle_peer_disconnect(data_connection){
+        console.log('a peer has been disconnected');
+        // remove the disconnected peer
+        this.session_peers = this.session_peers.filter(peerId => peerId !== data_connection.peer);
+        this.session_peers_conn = this.session_peers_conn.filter(conn => conn.peer !== data_connection.peer);
+    },
+
+    upload_document(new_data) {
+        this.doc_data.push(...new_data);
     },
 
     // Get document data from the tracker if no peer active for the document
     get_document_data(session_id) {
         this.tracker.emit('get_doc_data', session_id);
```

Comment:
If tracker is no longer handling an event called 'get_doc_data', we probably don't need this?

Reply:
That's right, I will remove it in the next set of changes.

Reply:
Removed in the lastest commit.